plugins/e57/io/E57Reader.cpp

```diff
@@ -51,30 +50,122 @@ std::string E57Reader::getName() const
 }
 
 E57Reader::E57Reader()
-    : Reader(), Streamable()
+    : Reader(), Streamable(), m_currentIndex(0), m_pointsInCurrentBatch(0), m_defaultChunkSize(1000000),
+      m_currentScan(-1)
+{
+}
+
+void E57Reader::addArgs(ProgramArgs& args)
 {
+    args.add("extra_dims", "Extra dimensions to write to E57 data",
+             m_extraDimsSpec);
+}
+
+void E57Reader::initializeBuffers()
+{
+    m_doubleBuffers.clear();
+    m_destBuffers.clear();
+    m_e57PointPrototype.reset(new StructureNode(m_scan->getPointPrototype()));
+
+    // Initialize for supported dimensions.
+    auto supportedFields = e57plugin::supportedE57Types();
+    for (auto& dimension : supportedFields)
+    {
+        if (m_e57PointPrototype->isDefined(dimension))
+        {
+            m_doubleBuffers[dimension] =
+                std::vector<double>(m_defaultChunkSize, 0);
+        }
+    }
+
+    // Initialize for extra dimensions.
+    for (auto i = m_extraDims->begin(); i != m_extraDims->end(); ++i)
+    {
+        if (m_e57PointPrototype->isDefined(i->m_name))
+        {
+            m_doubleBuffers[i->m_name] =
+                std::vector<double>(m_defaultChunkSize, 0);
+        }
+    }
+
+    // Link to destination buffers.
+    for (auto& keyValue : m_doubleBuffers)
+    {
+        m_destBuffers.emplace_back(
+            *m_imf, keyValue.first, keyValue.second.data(), m_defaultChunkSize,
+            true,
+            (m_e57PointPrototype->get(keyValue.first).type() ==
+             e57::E57_SCALED_INTEGER));
+    }
 }
 
 void E57Reader::addDimensions(PointLayoutPtr layout)
 {
-    for (auto& keyVal : m_doubleBuffers)
+    auto supportedFields = e57plugin::supportedE57Types();
+    auto numScans = m_data3D->childCount();
+    std::unique_ptr<Scan> tempScan;
+
+    for (auto& dimension : supportedFields)
```

Comment:
This is confusing to me.  Can you explain what the gist of it is?  The `extra_dims` refers to writing, but this is a reader.

Reply:
`extra_dims` can be used with both reader and writer. 

Consider a dimension `abcd` written by a writer.
Since the dimension is not in `supportedE57Types`, We also need to have support in reader to read that dimension. In that case the dimension need to be provided in `extra_dims`.
Here I am registering supported types which are present in E57 scan. And then same for extra dims.

Reply:
I believe the las reader has the same logic at the moment